ASoC: SOF: (re-)Introduce the foundation code for IPC4 support

ujfalusi opened the pull request:
Hi,

This series is intended to pave the way to add full support for IPC4 in SOF as it is re-introducing the code needed for handling IPC communication with the DSP running IPC4 version of a firmware.

Followup sets will be needed to add support for IPC4 specific firmware, topology, control, pcm and platform handling, but they all depend on this, basic and mandatory series.

Since the removal of the incomplete ipc4 implementation the include/sound/sof/ipc4/header.h has been changed quite a bit to make it easier to use (and to allow a bit more readable code).
The sound/soc/sof/ipc4.c file has been mostly rewritten and all functionality has been verified to be working (IPC tx, reply, rx and set_get_data), we also have different error handling and message logging. All external API is generic (no `ipc4` in their name).

Checkpatch is not happy about the `uint32_t`, but I was told that this is fine, so I have kept it like that.

Regards,
Peter

ranj063 commented on sound/soc/sof/trace.c:
```diff
@@ -450,6 +450,10 @@ int snd_sof_init_trace(struct snd_sof_dev *sdev)
 {
 	int ret;
 
+	/* dtrace is only supported with SOF_IPC */
```
nit-pick, typo in commit message for both the last 2 patches, missing "is" " Currently dtrace is only supported with SOF_IPC"

ranj063 replied:
btw, the last 2 patches are not meant for upstream right? It will be replaced with the correct version of trace and client code for each IPC version?

ujfalusi replied:
@ranj063, I have patches for the message injector to support IPC4, but the IPC flood afaik is IPC3 only, unless we bombard IPC4 firmware with something else (we have a dedicated message for this in IPC3).
The tracing support is different in IPC4, I don't think it makes any sense to try to force dma-trace or etrace on it. Likely we are going to have a different debugfs file for the tracing support.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -351,6 +351,9 @@ struct snd_sof_ipc_msg {
 
+	/* notification, firmware initiated messages */
+	void *rx_data;
+
```
commit message typos: 

its implementation.
cannot

ujfalusi replied:
Thanks, I have fixed it

plbossart commented on include/sound/sof/header.h:
```diff
@@ -132,6 +132,7 @@
 
 /* maximum message size for mailbox Tx/Rx */
 #define SOF_IPC_MSG_MAX_SIZE			384
+#define SOF_IPC4_MSG_MAX_SIZE			4096
```
could this go in an IPC4-specific header? All the stuff in this file is IPC3-specific.

ujfalusi replied:
Sure, I will move

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,438 @@
+	SOF_IPC4_GLB_RESTORE_PIPELINE,
+
+	/* Loads library (using Code Load or HD/A Host Output DMA) */
+	SOF_IPC4_GLB_LOAD_LIBRARY,
```
25 is missing? Should have a comment on discontinuities.

ujfalusi replied:
OK

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,438 @@
+	SOF_IPC4_MSB_INTEGER, /* integer with Most Significant Byte first */
+	SOF_IPC4_LSB_INTEGER, /* integer with Least Significant Byte first */
+	SOF_IPC4_SIGNED_INTEGER,
+	SOF_IPC4_UNSIGNED_INTEGER,
```
we should probably add something that describes what those two mean. I must admit not recalling why MSB/LSB and signed/unsigned would be incompatible mutually-exclusive concepts.

ujfalusi replied:
I could not fine any documentation on these apart from the fact that we are only using MSB_INTEGER.
I'll drop the last three entries and keep only the MSB and LSB type as they do make sense.
We can add back entries later if they ever surface since then we _know_ what they are.

The sample type is coming directly from topology.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+			       16, 4, ipc_data, size, false)
+#else
+#define sof_ipc4_dump_payload(sdev, ipc_data, size)	do { } while (0)
+#endif
```
why not just use print_hex_dump_debug without this DEBUG_VERBOSE?

ujfalusi replied:
True, I should have  used that version.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+	case 105 ... 109:
+	case 114 ... 115:
+	case 160 ... 163:
+	case 155:
```
155 is not in the list above? It's odd to have an 'unknown error' that would explicitly map as -ENOENT?

ujfalusi replied:
OK, I'll check.

ujfalusi replied:
155 should have been the 165, so typo.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+	case 11:
+	case 105 ... 109:
+	case 114 ... 115:
+	case 160 ... 163:
```
165 is missing here?

ujfalusi replied:
oh, I'll check it.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+		if (data_size_valid && msg->data_size)
+			dev_dbg(dev, "%s: %#x|%#x: %s|%s [data size: %zu]\n",
+				text, msg->primary, msg->extension, str, str2,
+	   msg->data_size);
```
alignment?

ujfalusi replied:
OK

ujfalusi commented:
Changes since v1:
- typos fixed
- Keep only document-able and  used values fro sample_type

ujfalusi commented:
> @ujfalusi the code was originally developed by Bard and Rander. Can we please add thei co-developed-by+signed-off-by in the relevant patches?

I will add the co-developed-by tag for sure, but I'm not confident to add their SoB as not much left from the original code and logic.

@bardliao, @RanderWang, what is your preference?

plbossart commented:
> > @ujfalusi the code was originally developed by Bard and Rander. Can we please add thei co-developed-by+signed-off-by in the relevant patches?
> 
> I will add the co-developed-by tag for sure, but I'm not confident to add their SoB as not much left from the original code and logic.
> 
> @bardliao, @RanderWang, what is your preference?

you don't have a choice @ujfalusi if you add the Co-developed-by tag the Signed-off-by tag needs to follow immediately.

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by

````
Since Co-developed-by: denotes authorship, every Co-developed-by: must be immediately followed by a Signed-off-by: of the associated co-author.
````

ujfalusi commented:
> > > @ujfalusi the code was originally developed by Bard and Rander. Can we please add thei co-developed-by+signed-off-by in the relevant patches?
> > 
> > 
> > I will add the co-developed-by tag for sure, but I'm not confident to add their SoB as not much left from the original code and logic.
> > @bardliao, @RanderWang, what is your preference?
> 
> you don't have a choice @ujfalusi if you add the Co-developed-by tag the Signed-off-by tag needs to follow immediately.
> 
> https://www.kernel.org/doc/html/latest/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by
> 
> ```
> Since Co-developed-by: denotes authorship, every Co-developed-by: must be immediately followed by a Signed-off-by: of the associated co-author.
> ```

You are right, both tag then.

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+		/* Module message */
+		switch (type) {
+		case SOF_IPC4_MOD_INIT_INSTANCE:
+			str = "MOD_INIT_INSTANCE";
```
case is very clear to show control flow but not efficient and the function code is a little large. Since type is a integer and continuous, how about to follow above fw status check: define a struct array with type and message string, so we can do it with "str = ipc_msg[type]" ?

ujfalusi replied:
@RanderWang, hm, let me see. I'll see how it looks.

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+		/* Module message */
+		switch (type) {
+		case SOF_IPC4_MOD_INIT_INSTANCE:
+			str = "MOD_INIT_INSTANCE";
```
case is very clear to show control flow but not efficient. Since type is a integer and continuous, how about to follow above fw status check: define a struct array with type and message string, so we can do it with "str = ipc_msg[type]" ?

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+	switch (status) {
+	case 8:
+	case 11:
+	case 105 ... 109:
```
another idea: can we integrate ret value to ipc4_status[i], then we can do it like
````
if (ipc4_status[i].status == status) {
			dev_err(sdev->dev, "FW reported error: %s\n", ipc4_status[i].msg);
			return ipc4_status[i].kernel_status;
}
````
it is easy to match fw status to kernel status

ujfalusi replied:
@RanderWang, I thought about that, but it would just adds dead weight (166 * sizeof(int) data) for not much benefit.

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+		return -EINVAL;
+
+	ret = sof_ipc_send_msg(sdev, msg_data, msg_bytes, reply_bytes);
+
```
remove the white line

ujfalusi replied:
OK.

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,669 @@
+		return;
+
+	ipc4_msg->data_ptr = NULL;
+	ipc4_msg->data_size = 0;
```
at the end of this function we will reset it. Do we need to do it at the beginning for each rx msg ?
````c
	if (data_size) {
		kfree(ipc4_msg->data_ptr);
		ipc4_msg->data_ptr = NULL;
		ipc4_msg->data_size = 0;
	}
````

ujfalusi replied:
I have added this to shield from lazy callers who failed to reset these (ipc4_msg allocated on stack w/o initialization).

ujfalusi commented:
Changes since v2:
- co-developed-by and signed-off-by tag fro @RanderWang and @bardliao added
- String table lookup for the message types as suggested by @RanderWang

ujfalusi commented:
checkpatch compains about the macros:
```
ERROR: space prohibited before open square bracket '['
#129: FILE: sound/soc/sof/ipc4.c:129:
+#define DBG_IPC4_MSG_TYPE_ENTRY(type)  [SOF_IPC4_##type] = #type
```
If I brake the macro to two lines, then checkpatch is happy:
```
#define DBG_IPC4_MSG_TYPE_ENTRY(type)	\
		[SOF_IPC4_##type] = #type
```
But I don't consider this to be a good 'fix', the macro works as intended and preprocessor spills out this:
```
static const char * const ipc4_dbg_mod_msg_type[] = {
 [SOF_IPC4_MOD_INIT_INSTANCE] = "MOD_INIT_INSTANCE",
 [SOF_IPC4_MOD_CONFIG_GET] = "MOD_CONFIG_GET",
 [SOF_IPC4_MOD_CONFIG_SET] = "MOD_CONFIG_SET",
 [SOF_IPC4_MOD_LARGE_CONFIG_GET] = "MOD_LARGE_CONFIG_GET",
 [SOF_IPC4_MOD_LARGE_CONFIG_SET] = "MOD_LARGE_CONFIG_SET",
 [SOF_IPC4_MOD_BIND] = "MOD_BIND",
 [SOF_IPC4_MOD_UNBIND] = "MOD_UNBIND",
 [SOF_IPC4_MOD_SET_DX] = "MOD_SET_DX",
 [SOF_IPC4_MOD_SET_D0IX] = "MOD_SET_D0IX",
 [SOF_IPC4_MOD_ENTER_MODULE_RESTORE] = "MOD_ENTER_MODULE_RESTORE",
 [SOF_IPC4_MOD_EXIT_MODULE_RESTORE] = "MOD_EXIT_MODULE_RESTORE",
 [SOF_IPC4_MOD_DELETE_INSTANCE] = "MOD_DELETE_INSTANCE",
};
```

RanderWang reviewed (approved):
LGTM
